rfd/0237-sub-ca-support.md

```diff
@@ -0,0 +1,623 @@
+    # OK, all overrides are addressed.
+    ```
+
+## Details
+
+The design works similarly to [RFD 0194 - SPIFFE issuer override][rfd0194]: we
+retain the Teleport generated private key but "override" the self-signed
+certificate with an externally-signed one. Retaining the private key and
+overriding only the self-signed certificate simplifies the implementation and
+allows overrides to take effect without a CA rotation.
+
+Unlike RFD 0194, the override is conceptually generic (it may apply to any
+Teleport CA) and works at a more fundamental system layer. For the initial
+release only the "db-client" and "windows" CAs may be targeted, but support
+could be seamlessly expanded in the future. It could even be applied to the
+SPIFFE TLS CA, replacing a workload_identity_x509_issuer_override.
+
+Overrides for client facing CAs, like "db-client" and "windows", take effect
+immediately. The customer holds all necessary certificates before creating the
+override, so they may take the necessary steps (like updating trusted roots).
+
+[rfd0194]: https://github.com/gravitational/teleport/blob/master/rfd/0194-spiffe-x509-override.md
+
+### The cert_authority_override resource
+
+A new Terraform-friendly resource is added to configure CA overrides. The new
+resource avoids direct changes to cert_authority, a sensitive Teleport-managed
+resource.
+
+Creating a new override automatically generates the corresponding empty CRL.
+
+```proto
+package teleport.subca.v1;
+
+message CertAuthorityOverride {
+  teleport.header.v1.ResourceHeader header = 1;
+  CertAuthorityOverrideSpec spec = 2;
+  CertAuthorityOverrideStatus status = 3;
+}
+
+message CertAuthorityOverrideSpec {
+  repeated CertificateOverride certificate_overrides = 1;
+}
+
+message CertificateOverride {
+  // SHA256 of the certificate's DER-encoded SubjectPublicKeyInfo
+  // (aka RawSubjectPublicKeyInfo).
+  // Informative if certificate is present.
+  string public_key = 1;
+
+  // Informative. Cluster of the CA.
+  string cluster_name = 2;
+
+  // Certificate to present, in PEM form.
+  //
+  // The public key must match an existing (CA,cluster) pair.
+  // It must also match the public_key field, if present.
+  //
+  // The Subject's "O=" field must match the CA cluster.
+  // It must also match the cluster_name field, if present.
+  string certificate = 3;
+
+  // Certificate chain, in PEM form.
+  //
+  // If present Teleport may supply the chain along with the certificate in
+  // appropriate situations.
+  repeated string chain = 4;
+
+  // TBD.
+  // bool exclude_sub_ca_from_client_chains = n;
+
+  // If true disables the override.
+  // A disabled override may exist simply to mark a certain public key as not
+  // overridden. In this case the certificate may be absent.
+  bool disabled = 5;
+}
+
+message CertAuthorityOverrideStatus {
+  map<string, CertificateRevocationList> public_key_to_crls = 1;
+}
+
+message CertificateRevocationList {
```

Comment:
I'm planning to move some of the base X.509 proto definitions to their own package. This is a start: https://github.com/gravitational/teleport/pull/61974.

This includes CRLs, DNs, etc.